### DISCUSSION TITLE
Two-stage keyboard shortcut support

### DISCUSSION THREAD

**Discussion by searls** [rxhanson/Rectangle#614] [Ideas]
I'm just checking out Rectangle because it looks like macOS Monterey has finally killed [Divvy](https://mizage.com/windivvy/).

I'd like to use Rectangle to replicate my Divvy keyboard configuration, which is a two-stage keyboard shortcuts:

1. Press `Ctrl-Space` to activate Divvy
2. Activate a windowing mode like:
        - `f` for full screen
        - `q` for left third
        - `w` for center third
        - `e` for right third

Is something like this possible with Rectangle? Assuming there's not a modal activator shortcut like with Divvy, is it possible that multi-stage keyboard shortcuts (e.g. `ctrl-space, <pause> f` might be configurable?

**Comment by rxhanson** [comment]
Thanks for checking out Rectangle. I have received this request before. From a developer perspective, this would require listening to keystrokes, which is not hard to add in but has some implications/complexity that are not desirable. Right now, the app just registers keyboard shortcuts. 

Since Rectangle has behavior that can occur by repeating commands, like pressing first third multiple times to cycle thirds, my subjective opinion is that I don’t personally see a huge benefit to your desired behavior other than that it fits what Divvy users might be used to. I could see considering this or considering a PR for it if with significant demand, though. 

As an aside, the custom shortcut builder in [Rectangle Pro](https://rectangleapp.com/pro) (paid) allows for more configuration of repeated commands.

**Comment by searls** [comment]
Thank you so much for the thorough reply!

**Comment by samholmes** [comment]
A modal experience would be nice because it would mean one wouldn't have to download another program like Karabiner or Hyperkey to tuck away the shortcuts to some difficult, unlikely to conflict keyboard binding.

The idea is to make a single keybinding to enter the Rectangle mode where simple keys like `hjkl` could move a window to one of the snap position and `n` and `p` keys could focus on the next/previous window on the desktop. This is just an idea for the bindings, not necessary the exact bindings we default to. A user configuration similar to `.vimrc` could allow the user to modify the keybindings to whatever they want. Of course, `Esc` would exit the snap mode.

This idea of "modal" UX isn't new; vim has been around for a while. By introducing a mode, the is no risk for keybinding conflicts within programs.

**Reply by rxhanson** [reply]
I understand, and although I don't have a personal need or desire for this, I would welcome a good pull request if anyone wanted to add it into Rectangle. The only catch is that Rectangle is meant to be extremely simple, so I would only want this to be enabled via Terminal commands, unless there is significant demand and a simple way to add it into the UI.

**Comment by rdavison** [comment]
Another Divvy refugee here... I would be glad to contribute a PR, but the truth is I don't have even the slightest clue how I would go about adding such a feature. I can probably replicate this modal behavior with Karabiner, but even then it's not quite the same. I'd be glad to pay for it if it were only added to Rectangle Pro. Btw, speaking of Rectangle Pro, the throwing feature is indeed really cool. I love the idea... currently playing around with it in the demo version, but ideally I'd prefer to avoid using the mouse for window management (Charmstone on the other hand is very tempting...). Then again, muscle memory is muscle memory. Gonna demo this feature for the next few days and see how I feel in a week.

**Reply by rxhanson** [reply]
Thanks for considering contributing and for looking at my other apps. A while has passed since I've given this thought, but there are more ways to achieve this, although they would probably just make more sense to add to Rectangle Pro, working in conjunction with what's already there. I recently added some foundation work there for keyboard shortcut behavior so I can make the window throw work using keyboard shortcuts (similar to Charmstone with the arrow keys), and that might be the simplest route to having this style of behavior (although not identical). 

With that said, if you decide you do want to add something to Rectangle, it's fairly straightforward to quickly get a grasp on which things do what. The WindowManager class is responsible for all of the logic of determining the size & position of a window, using calculation classes, and then it will execute the call to tell macOS to move/resize the window. The ShortcutManager class is responsible for receiving a notification that a window should be moved/resized, and it calls the WindowManager to do so.

**Reply by rdavison** [reply]
Thanks for your thorough response. If you do decide to go the Rectangle Pro route, that would certainly be a welcome addition, but if you don't find time for it, then I'll survive with a temporary Karabiner solution until maybe some day I can find the time to sit down and study the source code for Rectangle. 

I appreciate the hints you left in the second paragraph. They may come in handy some day.. but honestly I would need to familiarize myself with Swift and MacOS development, neither of which I have any real experience in. So it may be a while :)
